system-addon/content-src/components/TopSites/TopSiteForm.jsx

```diff
@@ -12,13 +12,35 @@ export class TopSiteForm extends React.PureComponent {
     this.state = {
       label: site ? (site.label || site.hostname) : "",
       url: site ? site.url : "",
-      validationError: false
+      validationError: false,
+      customScreenshotUrl: site ? site.customScreenshotURL : "",
+      showCustomScreenshotForm: site ? site.customScreenshotURL : false,
+      screenshotRequestFailed: false,
+      screenshotPreview: null,
+      pendingScreenshotUpdate: false
     };
+    this.onClearScreenshotInput = this.onClearScreenshotInput.bind(this);
     this.onLabelChange = this.onLabelChange.bind(this);
     this.onUrlChange = this.onUrlChange.bind(this);
     this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
     this.onClearUrlClick = this.onClearUrlClick.bind(this);
     this.onDoneButtonClick = this.onDoneButtonClick.bind(this);
+    this.onCustomScreenshotUrlChange = this.onCustomScreenshotUrlChange.bind(this);
+    this.onPreviewButtonClick = this.onPreviewButtonClick.bind(this);
+    this.onEnableScreenshotUrlForm = this.onEnableScreenshotUrlForm.bind(this);
+    this.validateUrl = this.validateUrl.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if ((nextProps.screenshotPreview && this.state.pendingScreenshotUpdate) ||
+        nextProps.screenshotRequestFailed) {
+      this.setState({
+        pendingScreenshotUpdate: false,
+        validationError: nextProps.screenshotRequestFailed,
+        screenshotRequestFailed: nextProps.screenshotRequestFailed,
+        screenshotPreview: nextProps.screenshotPreview
```

Comment:
The problem I was trying to work around here is that I want to use`props.screenshotRequestFailed` but if the user changes the value of the URL that prop value is obsolete and there is no way to clear it/know that it was used for a different request.

Reply:
Scenario
1. User has a custom screenshot
2. User previews a different screenshot and request fails
3. User resets to the initial custom url value
The button should say save, but because `props.screenshotRequestFailed` is true it will say "Preview" or prevent saving.
And there might be other scenarios (like user removes custom url completely). You could work around (and I tried) but I ended up having a lot of boolean operations that were not easy to reason about.

Reply:
Just noticed `this.state.screenshotRequestFailed` is not currently used to decide the previewMode as I described. Fixing this.